PS-5446: FLUSH CHANGED_PAGE_BITMAPS leaves gaps between the last written

gl-sergei opened the pull request:
bitmap LSN and InnoDB checkpoint LSN

Analysis.

xtrabackup determines the checkpoint lsn at the beginning of the backup
and starts the redo log copying. Then it invokes FLUSH
CHANGED_PAGE_BITMAPS.

FLUSH CHANGED_PAGE_BITMAPS implementation simply runs
log_online_follow_redo_log which may finish scanning before reaching
checkpoint lsn.

xtrabackup will than see that tracked lsn is less than checkpoint lsn
and changed page bitmaps are unusable.

Fix.

Fix is to save checkpoint lsn at the beginning of the flush
command. This lsn will be greater or equal to the one xtrabackup
has. Then run log_online_follow_redo_log until tracked lsn is greater or
equal to saved checkpoint lsn.

gl-sergei commented:
https://ps.cd.percona.com/view/8.0/job/percona-server-8.0-param/250/

laurynas-biveinis commented on storage/innobase/log/log0online.cc:
```diff
@@ -1350,13 +1350,34 @@ bool log_online_follow_redo_log(void) {
+
+  while (result && srv_track_changed_pages &&
+         last_checkpoint_lsn > log_sys->tracked_lsn.load()) {
+    result = log_online_follow_redo_log();
```
if (!result) break?

gl-sergei replied:
`if (!result) break'` is not needed, because the loop is `while (result)`

laurynas-biveinis replied:
Right, sorry

laurynas-biveinis commented on storage/innobase/include/log0online.h:
```diff
@@ -54,6 +54,13 @@ changed page bitmap which is then written to disk.
+
+@return true if log tracking succeeded, false if bitmap write I/O error
+*/
+bool log_online_follow_redo_log_past_checkpoint_lsn(void);
```
This is the main entry point into the changed page tracking - IMHO it should have a short name, such as original log_online_follow_redo_log, and that one should be renamed instead (i.e. log_online_follow_redo_log_one_pass or something else.)

gl-sergei replied:
makes sense, will rename

laurynas-biveinis commented on storage/innobase/include/log0online.h:
```diff
@@ -54,6 +54,13 @@ changed page bitmap which is then written to disk.
 @return true if log tracking succeeded, false if bitmap write I/O error */
 bool log_online_follow_redo_log(void);
```
Is it now feasible to mark this one MY_NODISCARD?

laurynas-biveinis commented on storage/innobase/log/log0online.cc:
```diff
@@ -1317,13 +1317,11 @@ static void log_online_parse_complete_recs_past_previous_checkpoint() noexcept {
 
 @return true if log tracking succeeded, false if bitmap write I/O error */
-bool log_online_follow_redo_log(void) {
+bool log_online_follow_redo_log_one_pass(void) {
```
Also, can this one be made static?

gl-sergei replied:
`srv_redo_log_follow_thread` is using `...one_pass` and I did not want change it. My original idea was to only change FLUSH command....

laurynas-biveinis replied:
Sure, no need to change that

gl-sergei commented:
Updated:
- `log_online_follow_redo_log renamed` to `log_online_follow_redo_log_one_pass`, added MY_NODISCARD

https://ps.cd.percona.com/view/8.0/job/percona-server-8.0-param/251/

percona_changed_page_bmp_log_resize failed again. Note, previous failure was in different PR.
